### PR TITLE
CFM-443: Create function to retrieve all user groups grouped by membership

### PR DESCRIPTION
**Pull request by DavidHernandez**
Refactorizes the way the user groups are loaded to meet the requested specification.

Issue: https://issuetracker.amplexor.com/jira/browse/CFM-443

No screenshot, just code changes.

### CODE REVIEW

**Comment by DavidHernandez** [comment]
## Depends on https://github.com/capacity4dev/capacity4more/pull/1138

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 1]
```diff
@@ -62,20 +46,90 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+
+  return $groups;
+}
+
+/**
+ * Returns the list of groups of an user, grouped by the membership type.
+ *
+ * @param object $account.
+ *   User object. Defaults global user if NULL.
+ * @param $types.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: `$ignored_types`

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 2]
```diff
@@ -62,20 +46,90 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+function c4m_features_homepage_get_user_groups_grouped_by_membership($account = NULL, $types = array('group')) {
+  $groups = c4m_features_homepage_get_user_groups($account, $types);
+
+  foreach ($groups as $group) {
+    // Ignores not wanted groups.
+    if (in_array($group->type, $types)) {
+      continue;
+    }
+
+    $group_admins = c4m_og_get_group_admins($group->nid);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: let's use `c4m_user_og_get_role_in_group()`

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 3]
```diff
@@ -62,20 +44,77 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+ *
+ * @param object $account.
+ *   User object. Defaults global user if NULL.
+ * @param $allowed_types.
+ *   Array of group types to be returned.
+ *
+ * @returns mixed|null
+ *   List of groups or NULL if the user has no groups.
+ */
+function c4m_features_homepage_get_user_groups($account = NULL, $allowed_types = array('group')) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @DavidHernandez 
why not using `c4m_user_og_get_groups()` instead ?

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 4]
```diff
@@ -62,20 +44,71 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+  if (is_null($account)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: I believe you can remove this one...

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 5]
```diff
@@ -62,20 +44,71 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+    global $user;
+    $account = user_load($user->uid);
+  }
+
+  $groups = array();
+  foreach ($allowed_types as $type) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @DavidHernandez 
do we still need this foreach ?
seems like we don't ...

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 6]
```diff
@@ -62,20 +44,71 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+    $groups += c4m_user_og_get_groups($account);
+  }
+
+  if (empty($groups)) {
+    return FALSE;
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: why not just returns an empty array ?

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 7]
```diff
@@ -70,7 +70,7 @@ function c4m_features_homepage_get_user_groups($account = NULL, $allowed_types =
   $groups = c4m_user_og_get_groups($account);
 
   if (empty($groups)) {
-    return FALSE;
+    return array();
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @DavidHernandez 
What's happening when using `usort` on an empty array ?

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 8]
```diff
@@ -21,7 +21,7 @@ $plugin = array(
 function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $args, $context) {
   $groups = c4m_features_homepage_get_user_groups();
 
-  if (is_null($groups)) {
+  if (empty($groups)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]:
```php
if (!$groups) {
```

should also work in here ...

Comment (DavidHernandez) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: Yes, that would also work, but the code would not be explicit. $groups is always an array, using !$groups could be understood as if $groups is a boolean.

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc (hunk 7 above)
Comment (DavidHernandez) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: I had to test it, I read first that $usort converts empty arrays to NULL, for that reason I didn't remove it before. But seems that it works normally, if receives an empty array, returns an empty array.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: might be dependent on the php version ?

Comment (DavidHernandez) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: No, usort implementation has not changed.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: so we can remove this condition and just `return $groups;` at the bottom

Comment (DavidHernandez) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: Already did that change: https://github.com/capacity4dev/capacity4more/pull/1139/files#diff-1056a588a4b868e01a197b176b661a81R71

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: sorry, didn't notice ... ;)

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc (hunk 3 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @DavidHernandez 
can we remove the `$allowed_types` ?

Comment (DavidHernandez) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @ordavidil no, that was requested on the issue.

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: but currently it is unused ...

File: project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc [hunk 9]
```diff
@@ -62,20 +44,54 @@ function c4m_features_homepage_my_groups_content_type_render($subtype, $conf, $a
...
+ * Returns the list of groups of an user, grouped by the membership type.
+ *
+ * @param object $account.
+ *   User object. Defaults global user if NULL.
+ *
+ * @returns mixed|null
+ *   List of groups or NULL if the user has no groups.
+ */
+function c4m_features_homepage_get_user_groups_grouped_by_membership($account = NULL) {
+  $groups = c4m_features_homepage_get_user_groups($account, $allowed_types);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features/c4m_features_homepage/plugins/content_types/my_groups/my_groups.inc]: @DavidHernandez 
we probably want to remove the `$allowed_types` from here too ...

**Comment by soniCaH** [comment]
@DavidHernandez #1138 says to merge this one first, but here it states the other way around. Could you please update either of both so I know which to merge first please? :-)

**Comment by ordavidil** [comment]
@soniCaH 
He meant that basically, you should review both of the PRs before any merge, and if everything goes fine then merge this one first, since it's compared to `feature/CFM-760-my-groups-block` branch, at that point that branch will include both of the PRs so you can merge #1138, since it is the one that is compared to `develop`

**Pull request merged**